### PR TITLE
Add TTFT benchmarks + update sparsity benchmarks

### PR DESCRIPTION
**Pull request by jcaip**
This PR adds in TTFT token benchmarks to torchAO, and also updates the benchmarking script to handle sparsity a bit nicer + use the 2:4 sparse checkpoints that are available. 

image

Additionally also adds in padding support for int8 dynamic quant + 2:4 sparsity, which we were missing before.

### CODE REVIEW

File: torchao/_models/llama/generate.py [hunk 1]
```diff
@@ -217,24 +232,29 @@ def main(
...
             else:
                 use_hqq=False
             groupsize=int(quantization.split("-")[1])
             assert groupsize in [32,64,128,256], f"int4wo groupsize needs to be one of [32,64,128,256] but got {groupsize}"
-            quantize_(model, int4_weight_only(group_size=groupsize))
-        if "marlin" in quantization:
-            from torchao.dtypes import MarlinSparseLayout
-            quantize_(model, int4_weight_only(layout=MarlinSparseLayout()))
+            if sparsity and "semi" in sparsity:
+                quantize_(model, int4_weight_only(layout=MarlinSparseLayout()))
```
Comment (HDCharles) [inline torchao/_models/llama/generate.py:255]: this isn't using any of the derived variables. It should use the derived ones or be in a separate section.

**Review by HDCharles — APPROVED** [review]
lgtm if you move the marlin stuff so its clearer what derived variables it actually uses

File: torchao/_models/llama/generate.py [hunk 2]
```diff
@@ -417,23 +446,28 @@ def callback(x):
...
     mem = torch.cuda.max_memory_reserved() /1e9
     print(f"Average tokens/sec: {tokpersec:.2f}")
     print(f"Average Bandwidth: {bandwidth:.02f} GB/s")
+    print(f"Average time: {avg_time:.04f} s")
     print(f"Peak Memory Usage: {mem:.02f} GB")
     print(f"Model Size: {model_size:.02f} GB")
     if write_result:
-        result_txt = f"\n{datetime.today().strftime('%Y%m%d%H%M%S')}, tok/s={tokpersec:6.2f}, mem/s={bandwidth:7.2f} GB/s, peak_mem={mem:5.2f} GB, model_size={model_size:5.2f} GB "
-        result_txt += f"quant: {quantization}, mod: {checkpoint_path.parent.name}, kv_quant: {kv_cache_quantization}, compile: {compile}, compile_prefill: {compile_prefill}, dtype: {precision}, device: {device} "
+        result_txt = f"\n{datetime.today().strftime('%Y%m%d%H%M%S')}, tok/s={tokpersec:6.2f}, mem/s={bandwidth:7.2f} GB/s, time={t:5.4f} sec, peak_mem={mem:5.2f} GB, model_size={model_size:5.2f} GB "
```
[the remark this answers is not in the document]
Comment (jcaip) [reply torchao/_models/llama/generate.py:460]: It's overall time, but I limit num_tokens to 1. I can make this a bit clearer though, maybe a --ttft flag that sets forces num_tokens_generated to be 1.

**Comment by zhyncs** [comment]
Hi @vkuzo Thanks for the great work!
I see that GitHub has already released 0.7.0, but PyPI is still at 0.6.1. When is the update for PyPI expected? Thanks!
https://github.com/pytorch/ao/releases/tag/v0.7.0

**Comment by vkuzo** [comment]
> Hi @vkuzo Thanks for the great work!
> 
> I see that GitHub has already released 0.7.0, but PyPI is still at 0.6.1. When is the update for PyPI expected? Thanks!
> 
> https://github.com/pytorch/ao/releases/tag/v0.7.0

It's available as of last night!

**Pull request merged**